datafusion/physical-optimizer/src/enforce_sorting/sort_pushdown.rs

```diff
@@ -216,7 +218,28 @@ fn pushdown_sorts_helper(
 fn pushdown_requirement_to_children(
     plan: &Arc<dyn ExecutionPlan>,
     parent_required: OrderingRequirements,
+    parent_fetch: Option<usize>,
 ) -> Result<Option<Vec<Option<OrderingRequirements>>>> {
+    // Only attempt to push down TopK when there is an upstream LIMIT
+    if parent_fetch.is_some() {
+        // 1) Never push a new TopK below an operator that already has its own fetch
+        if plan.fetch().is_some() {
+            return Ok(None);
+        }
+        // 2) Only allow pushdown through operators that do not increase row count
+        //    (equal or lower-equal cardinality). Any other operator (including joins,
+        //    sort-with-limit, or UDTFs that may expand rows) must stop the pushdown.
+        let effect = plan.cardinality_effect();
+        if !matches!(
+            effect,
+            CardinalityEffect::Equal | CardinalityEffect::LowerEqual
+        ) {
+            return Ok(None);
+        }
```

Comment:
Should this be


```suggestion
        let effect = plan.cardinality_effect();
        if !matches!(
            plan.cardinality_effect(),
            CardinalityEffect::Equal
        ) {
            return Ok(None);
        }
```

Reply:
Good suggestion! @adriangb 

I was adding CardinalityEffect::LowerEqual, because filterexec is LowerEqual, it should also not passing sort pushdown.

Reply:
Hmm but if `FilterExec` is `LowerEqual` aren't you saying "if the cardinality effect is not one of Equal or LowerEqual` allow pushdown" which is equivalent to "if the cardinality effect is either of Equal or LowerEqual do allow pushdown", which would be the opposite of what we want? I think we *only* want to allow pushdown if the cardinality effect is `Equal`

Reply:
But maybe I'm confused about what `plan` is referring to here, etc.

Reply:
I may add more testing cases as follow-up.

Reply:
Thank you @adriangb , i think i got it now, even filter exec, we should not pass the push down.

Changed it to equal in latest PR.